test/e2e/nodepool_machineconfig_test.go

```diff
@@ -0,0 +1,199 @@
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilpointer "k8s.io/utils/pointer"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/yaml"
+)
+
+func TestNodepoolMachineconfigGetsRolledout(t *testing.T) {
+	t.Parallel()
+	g := NewWithT(t)
+
+	ctx, cancel := context.WithCancel(testContext)
+	defer cancel()
+
+	client, err := e2eutil.GetClient()
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get k8s client")
+
+	clusterOpts := globalOpts.DefaultClusterOptions()
+	clusterOpts.ControlPlaneAvailabilityPolicy = string(hyperv1.SingleReplica)
+	clusterOpts.BeforeApply = func(o crclient.Object) {
+		nodePool, isNodepool := o.(*hyperv1.NodePool)
+		if !isNodepool {
+			return
+		}
+		nodePool.Spec.Management.Replace.RollingUpdate.MaxSurge = func(v intstr.IntOrString) *intstr.IntOrString { return &v }(intstr.FromInt(int(*nodePool.Spec.Replicas)))
+		nodePool.Spec.Management.Replace.RollingUpdate.MaxUnavailable = func(v intstr.IntOrString) *intstr.IntOrString { return &v }(intstr.FromInt(0))
+	}
+
+	hostedCluster := e2eutil.CreateCluster(t, ctx, client, &clusterOpts, globalOpts.Platform, globalOpts.ArtifactDir)
+
+	// Sanity check the cluster by waiting for the nodes to report ready
+	t.Logf("Waiting for guest client to become available")
+	guestClient := e2eutil.WaitForGuestClient(t, testContext, client, hostedCluster)
+
+	// Wait for Nodes to be Ready
+	numNodes := int32(globalOpts.configurableClusterOptions.NodePoolReplicas * len(clusterOpts.AWSPlatform.Zones))
+	e2eutil.WaitForNReadyNodes(t, testContext, guestClient, numNodes)
+
+	// Wait for the rollout to be complete
+	t.Logf("Waiting for cluster rollout. Image: %s", globalOpts.LatestReleaseImage)
+	e2eutil.WaitForImageRollout(t, testContext, client, hostedCluster, globalOpts.LatestReleaseImage)
+	err = client.Get(testContext, crclient.ObjectKeyFromObject(hostedCluster), hostedCluster)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get hostedcluster")
+
+	ignitionConfig := ignitionapi.Config{
+		Ignition: ignitionapi.Ignition{
+			Version: "3.2.0",
+		},
+		Storage: ignitionapi.Storage{
+			Files: []ignitionapi.File{{
+				Node:          ignitionapi.Node{Path: "/etc/custom-config"},
+				FileEmbedded1: ignitionapi.FileEmbedded1{Contents: ignitionapi.Resource{Source: utilpointer.String("data:,content%0A")}},
+			}},
+		},
+	}
+	serializedIgnitionConfig, err := json.Marshal(ignitionConfig)
+	if err != nil {
+		t.Fatalf("failed to serialize ignition config: %v", err)
+	}
+	machineConfig := &mcfgv1.MachineConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   "custom",
+			Labels: map[string]string{"machineconfiguration.openshift.io/role": "worker"},
```

Comment:
note aside from this PR: This is a necessary impl detail for our machine config design to work. We should be enforcing it in the controller rather than making the user set it.

Reply:
We can drop this now that you added https://github.com/openshift/hypershift/pull/1428, can do follow up PR if you prefer